### DISCUSSION TITLE
Restart after Server Error

### DISCUSSION THREAD

**Discussion by JMarstrell** [zwave-js/ZWaveJS.NET#26] [General]
Is there a way to restart the server after an error without closing and restarting my c# application? If not, is there any other way to recover from a ZWaveJS.NET server error?

**Comment by marcus-j-davies** [comment]
Hi @JMarstrell 

I do believe I implemented a  `Destroy()` method on the Driver class.
https://github.com/zwave-js/ZWaveJS.NET/blob/ffb4f89a15ae97ad5baf7aa262800bdbde1cef6a/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/Driver.cs#L438

Using this will bring it down, and you will need to fire up a new instance of this class to start over.
Its advisable to check any zombie processes after this  - but i'm pretty sure they are terminated.

I need to bring this package up to the latest API level, so any input is welcome

**Reply by JMarstrell** [reply]
The driver does go through the destroy process. Any attempt to destroy it again after fatal error gives an error.

If I try to create another instance of the driver after a fatal error the server doesn’t fully complete its startup. Not sure if the port 50001 is being released or not. I have been trying to find a way to check.

Thinking about it, the server my be running from a previous crash and an attempt to start the driver again works momentarily and then causes the fatal error. Is there a way to check and stop it or reconnect?

**Reply by marcus-j-davies** [reply]
Could you provide me a stack trace of a fatal error?

A fatal error should only occur during driver start up.

https://github.com/zwave-js/ZWaveJS.NET/blob/ffb4f89a15ae97ad5baf7aa262800bdbde1cef6a/PSI/server.js#L43

These are handled later here, where a fatal error event is emitted for the frontend app to restart everything (but not before calling `Destroy`)
https://github.com/zwave-js/ZWaveJS.NET/blob/ffb4f89a15ae97ad5baf7aa262800bdbde1cef6a/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/Driver.cs#L517

The `Terminate`  method (in the `Destroy` block) actually sends a message to the runtime, to destroy its own self

https://github.com/zwave-js/ZWaveJS.NET/blob/ffb4f89a15ae97ad5baf7aa262800bdbde1cef6a/Visual%20Studio%20Projects/ZWaveJS.NET/ZWaveJS.NET/Server.cs#L17

https://github.com/zwave-js/ZWaveJS.NET/blob/ffb4f89a15ae97ad5baf7aa262800bdbde1cef6a/PSI/server.js#L47



You could go for broke and kill the image

`Process.GetProcessesByName("Server.psi").kill()`
